sklearn/metrics/_dist_metrics.pyx.tp

```diff
@@ -507,31 +845,39 @@ cdef class SEuclideanDistance{{name_suffix}}(DistanceMetric{{name_suffix}}):
        D(x, y) = \sqrt{ \sum_i \frac{ (x_i - y_i) ^ 2}{V_i} }
     """
     def __init__(self, V):
-        self.vec = ReadonlyArrayWrapper(np.asarray(V, dtype={{DTYPE}}))
+        self.vec = ReadonlyArrayWrapper(np.asarray(V, dtype=DTYPE))
         self.size = self.vec.shape[0]
         self.p = 2
 
     def _validate_data(self, X):
         if X.shape[1] != self.size:
             raise ValueError('SEuclidean dist: size of V does not match')
 
-    cdef inline DTYPE_t rdist(self, const {{DTYPE_t}}* x1, const {{DTYPE_t}}* x2,
-                              ITYPE_t size) nogil except -1:
+    cdef inline DTYPE_t rdist(
+        self,
+        const {{INPUT_DTYPE_t}}* x1,
+        const {{INPUT_DTYPE_t}}* x2,
+        ITYPE_t size,
+    ) nogil except -1:
         cdef DTYPE_t tmp, d=0
         cdef cnp.intp_t j
         for j in range(size):
-            tmp = <DTYPE_t> (x1[j] - x2[j])
-            d += <DTYPE_t> (tmp * tmp / self.vec[j])
+            tmp = (<DTYPE_t>(x1[j]) - <DTYPE_t>(x2[j]))
+            d += (tmp * tmp / self.vec[j])
```

Comment:
This is the new up-casting strategy (upcasting before the difference, rather than after) that is likely the cause of the change required in `test_distance_metrics_dtype_consistency`.